### ISSUE TITLE
Behaviour of writing to [[SentDatagrams]] when state is not "connected" is undocumented

### ISSUE THREAD

**Issue by ricea** [w3c/webtransport#93]
I can see 3 possible behaviours:

1. Silently drop any datagrams when state is not "connected"
2. Error the stream if write() is called when the state is "connecting", "closed", or "failed"
3. Queue datagrams in "connecting" state, error the stream in "closed" or "failed".

Option 2. seems user-unfriendly, since failure or success will depend on timing. Option 1. is the easiest to implement.

**Comment by lgrahl** [comment]
2 would be my preferred choice as an application developer and as an implementer. Why do you think it is user-unfriendly?

I've made a suggestion in #92 regarding on how to avoid the `connecting` state issue.

**Comment by ricea** [comment]
> Option 2 would be my preferred choice as an application developer and as an implementer. Explicit errors are great. Why do you think it is user-unfriendly?

Suppose a developer writes:

```javascript
const quicTransport = new QuicTransport(url);
const datagram = await someAsyncWork();
const writer = quicTransport.sendDatagrams().getWriter();
await writer.write(datagram);
```
They find this works on their dev box, sometimes throws an exception when deployed to production, and for some customers never works at all.

Maybe this is still better than option 1., which also fails but doesn't produce any diagnostics.

**Comment by murillo128** [comment]
```
const writer = quicTransport.sendDatagrams().getWriter();
```
seems web transport does not define how/when to initialize the `[[writer]]` slot in the ` [[SentDatagrams]]`

according to the whatwg:
```
[[writer]] | A WritableStreamDefaultWriter instance, if the stream is locked to a writer, or undefined if it is not
```

so imho, (from a quick look) I would say that if the `[[writer]]` exist, it should not fail (if close/abort have not been called). 

Another option would be to not set the `[[writer]]` of the `[[SentDatagrams]]` until the transport is in connected state.

**Comment by lgrahl** [comment]
I think the example was meant to throw at line 4, so `await writer.write(datagram)`.

**Comment by yutakahirano** [comment]
I don't like 3. Given we don't queue datagrams when connected, queuing datagram(s) while connecting sounds weird.

**Comment by murillo128** [comment]
> I don't like 3. Given we don't queue datagrams when connected, queuing datagram(s) while connecting sounds weird.

Aren't datagrams supposed to be queued internally due to backpressure/congestion control?

**Comment by ricea** [comment]
> Aren't datagrams supposed to be queued internally due to backpressure/congestion control?

I believe they are dropped if they can't be sent immediately, as is expected for an unreliable transport.

Backpressure only applies to the unidirectional and bidirectional streams.

**Comment by Atrius** [comment]
No queuing of datagrams at all seems unlikely to be viable.  @vasilvv and I discussed this point last week, and thought it would be difficult or impossible to achieve reasonable throughput without queuing datagrams that are blocked by congestion control.

If we go that route, option 3 seems pretty reasonable to me.

As far as I can tell, the main difference between options 2 and 3 is that option 3 lets the application prepare datagrams which the transport can send immediately, as soon as it's connected, without having to wait for an async call to notify the application that it's connected.

If that's allowed with streams, I think it should be allowed with datagrams.

**Comment by yutakahirano** [comment]
> No queuing of datagrams at all seems unlikely to be viable. @vasilvv and I discussed this point last week, and thought it would be difficult or impossible to achieve reasonable throughput without queuing datagrams that are blocked by congestion control.

Thank you. But even in that case we will have a size-bounded queue and datagarams will be discarded when the queue is full, right?

**Comment by vasilvv** [comment]
> Thank you. But even in that case we will have a size-bounded queue and datagarams will be discarded when the queue is full, right?

Indeed.  The current plan I have is to have a queue with a time limit (since a good time bound is probably easier to pick than a good size bound).

I need to file a separate issue to track this, but I just want to document the motivation for this before I forget.  If a QUIC stack uses a purely window based congestion control, datagrams work roughly as most users expect: there's a certain safe number of datagrams one can send per round-trip (CWND), and all data exceeding that limit is dropped.  The drop usually happens to the datagrams at the end of a flight, and a lot of algorithms optimize for that fact.

Google QUIC uses rate-based congestion control and pacing; the IETF draft does recommend using pacing, so I'd expect other implementations to have the same problem.  When pacing is used, the congestion controller picks a target bandwidth, and only allows `target bandwidth * 1ms` within every 1ms period.  Dropping packets immediately would force the users of this API to try to align the datagram writes with the 1ms timer in the network process; that would probably work poorly and consume a lot of IPC bandwidth.

**Comment by ricea** [comment]
Looking at #152 I realised it doesn't make much sense for `transport.datagramWritable.getWriter().ready` to resolve before the connection is complete. So maybe the solution is for the underlying sink not to resolve the promise returned by [startAlgorithm ](https://streams.spec.whatwg.org/#create-writable-stream) until connection succeeds. This implies the queuing behaviour described in 3., but backpressure will be applied until connection succeeds.

**Comment by jan-ivar** [comment]
Sounds like we're converging on 3, which I agree with.
